src/ess/reduce/nexus/generic_types.py

```diff
@@ -0,0 +1,169 @@
+Monitor3 = NewType('Monitor3', int)
+"""Identifier for an arbitrary monitor"""
+Monitor4 = NewType('Monitor4', int)
+"""Identifier for an arbitrary monitor"""
+Monitor5 = NewType('Monitor5', int)
+"""Identifier for an arbitrary monitor"""
+Incident = NewType('Incident', int)
+"""Incident monitor"""
+Transmission = NewType('Transmission', int)
+"""Transmission monitor"""
+MonitorType = TypeVar(
+    'MonitorType',
+    Monitor1,
+    Monitor2,
+    Monitor3,
+    Monitor4,
+    Monitor5,
+    Incident,
+    Transmission,
+)
+"""TypeVar used for specifying Incident or Transmission monitor type"""
+
+
+class NeXusMonitorName(sciline.Scope[MonitorType, str], str):
+    """Name of a monitor in a NeXus file."""
+
+
+class NeXusDetector(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Full raw data from a NeXus detector."""
+
+
+class NeXusMonitor(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataGroup], sc.DataGroup
+):
+    """Full raw data from a NeXus monitor."""
+
+
+class NeXusSample(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Raw data from a NeXus sample."""
+
+
+class NeXusSource(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Raw data from a NeXus source."""
+
+
+class NeXusDetectorEventData(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Data array loaded from a NeXus NXevent_data group within an NXdetector."""
+
+
+class NeXusMonitorEventData(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Data array loaded from a NeXus NXevent_data group within an NXmonitor."""
+
+
+class SourcePosition(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Position of the neutron source."""
+
+
+class SamplePosition(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Position of the sample."""
+
+
+class DetectorPositionOffset(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Offset for the detector position, added to base position."""
+
+
+class MonitorPositionOffset(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.Variable], sc.Variable
+):
+    """Offset for the monitor position, added to base position."""
+
+
+class CalibratedDetector(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Calibrated data from a detector."""
+
+
+class CalibratedMonitor(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Calibrated data from a monitor."""
+
+
+class DetectorData(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Calibrated detector merged with neutron event data."""
+
+
+class MonitorData(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Calibrated monitor merged with neutron event data."""
+
+
+Component = TypeVar('Component', bound=snx.NXobject)
+
+
+@dataclass
+class NeXusFileSpec(Generic[RunType]):
+    value: FilePath | NeXusFile | NeXusGroup
+
+
+@dataclass
+class NeXusComponentLocationSpec(
```

Comment:
Look closely:
```python
@dataclass
class NeXusComponentLocationSpec(
    NeXusLocationSpec[Component], Generic[Component, RunType]
): ...

@dataclass
class NeXusMonitorLocationSpec(
    NeXusLocationSpec[snx.NXmonitor], Generic[RunType, MonitorType]
): ...
```
`NeXusComponentLocationSpec` is generic in the *component* and is used also for sample and source. We cannot use `NeXusComponentLocationSpec[snx.NXmonitor, RunType]` since monitors need one addition type argument (`MonitorType`), so there is the special `NeXusMonitorLocationSpec`.

Reply:
Ah, I missed that it was also used for sample and source.